relayer/middleware/metrics.middleware.ts

```diff
@@ -1,45 +1,173 @@
+  vaaTimestamp(): number {
+    // vaa.timestamp is in seconds
+    return this.job?.data?.parsedVaa?.timestamp * 1000 ?? 0;
+  }
+
+  processingTime(): number {
+    return this.endTime - this.startTime;
+  }
+
+  totalTime(): number {
+    if (this.vaaTimestamp() > 0) {
+      return this.endTime - this.vaaTimestamp();
+    }
+
+    return 0;
+  }
+}
+
+const status = "status";
+const terminal = "terminal";
+
+export interface MetricsOpts<C extends StorageContext> {
   registry?: Registry;
+  labels?: MetricLabelsOpts<C>;
+  processingTimeBuckets?: number[];
+}
+
+export class MetricLabelsOpts<C extends StorageContext> {
+  labelNames: string[];
+  customizer: (context: C) => Promise<MetricRecord>;
+
+  constructor(
+    labelNames: string[] = [],
+    customizer: (ctx: C) => Promise<MetricRecord>,
+  ) {
+    this.labelNames = labelNames;
+    this.customizer = customizer;
+  }
 }
 
-export function metrics(
-  opts: MetricsOpts = {},
-): Middleware<Context & { job?: Job }> {
+const getLabels = async <C extends StorageContext>(
+  ctx: C,
+  opts: MetricsOpts<C>,
+  failed: boolean,
+) => {
+  const defaultLabels = { status: failed ? "failed" : "succeeded" };
+
+  try {
+    const labels: MetricRecord = await (opts.labels?.customizer(ctx) ??
+      Promise.resolve({}));
+
+    const allowedKeys = opts.labels?.labelNames ?? [];
+    const validatedLabels: MetricRecord = {};
+    allowedKeys.forEach(key => {
+      if (labels[key]) {
+        validatedLabels[key] = labels[key];
+      }
+    });
+
+    return { ...defaultLabels, ...validatedLabels };
+  } catch (error) {
+    ctx.logger?.error(
+      "Failed to customize metric labels, please review",
+      error,
+    );
+
+    return defaultLabels;
+  }
+};
+
+export function metrics<C extends StorageContext>(
+  opts: MetricsOpts<C> = {},
+): Middleware<C> {
   opts.registry = opts.registry || register;
+
   const processedVaasTotal = new Counter({
     name: "vaas_processed_total",
-    help: "Number of vaas processed successfully or unsuccessfully.",
+    help: "Number of incoming vaas processed successfully or unsuccessfully.",
+    labelNames: [].concat(opts.labels?.labelNames ?? []),
     registers: [opts.registry],
   });
 
   const finishedVaasTotal = new Counter({
     name: "vaas_finished_total",
     help: "Number of vaas processed successfully or unsuccessfully.",
-    labelNames: ["status"],
+    labelNames: [status, terminal].concat(opts.labels?.labelNames ?? []),
     registers: [opts.registry],
   });
 
-  return async (ctx: StorageContext, next) => {
-    const job = ctx.storage?.job;
-    // disable this metric if storage is enabled because the storage will actually compute the metrics.
-    if (job) {
-      await next();
-      return;
-    }
+  const processingDuration = new Histogram({
+    name: "vaas_processing_duration",
+    help: "Processing time in ms for jobs",
+    buckets: opts.processingTimeBuckets ?? [
+      6000, 7000, 7500, 8000, 8500, 9000, 10000, 12000,
```

Comment:
Would be better if we can extract this arrays of numbers as constant and re-use like `defaultTimeBuckets`